Sources/NIOFileSystem/Internal/ParallelDirCopy.swift

```diff
@@ -0,0 +1,158 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(macOS) || os(iOS) || os(tvOS) || os(watchOS) || os(Linux) || os(Android)
+import NIOCore
+
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+extension FileSystem {
+
+    /// Iterative implementation of a recursive parallel copy of the directory from `sourcePath` to `destinationPath`.
+    ///
+    /// The parallelism is solely at the level of individual items (so files, symbolic links and directories), a larger file
+    /// is not considered for being 'split' into concurrent reads or wites.
+    /// If any symbolic link is encountered then only the link is copied.
+    /// The copied items will preserve permissions and any extended attributes (if supported by the file system).
+    ///
+    /// Note: `maxConcurrentOperations` is used as a hard (conservative) limit on the number of open file descriptors at any point.
+    /// Operations are assume to consume 2 descriptors so maximum open descriptors is `maxConcurrentOperations * 2`
+    @usableFromInline
+    func copyDirectoryParallel(
+        from sourcePath: FilePath,
+        to destinationPath: FilePath,
+        maxConcurrentOperations: Int,
+        shouldProceedAfterError: @escaping @Sendable (
+            _ entry: DirectoryEntry,
+            _ error: Error
+        ) async throws -> Void,
+        shouldCopyItem: @escaping @Sendable (
+            _ source: DirectoryEntry,
+            _ destination: FilePath
+        ) async -> Bool
+    ) async throws {
+        // Implemented with NIOAsyncSequenceProducer rather than AsyncStream.
+        // It is approximately the same speed in the best case but has significantly less variance.
+        // NIOAsyncSequenceProducer also enforces a multi producer single consumer access pattern.
+        let sequence = NIOAsyncSequenceProducer<DirCopyItem, NoBackPressureStrategy, DirCopyDelegate>.makeSequence(
+            backPressureStrategy: .init(),
+            finishOnDeinit: false,
+            delegate: .init()
+        )
```

Comment:
Can we re-spell this slightly for readability?

```suggestion
        let sequence = NIOAsyncSequenceProducer.makeSequence(
            elementType: DirCopyItem.self,
            backPressureStrategy: NoBackPressureStrategy(),
            finishOnDeinit: false,
            delegate: DirCopyDelegate()
        )
```